metagraph/src/annotation/binary_matrix/row_sparse_disk/row_sparse_disk.cpp

```diff
@@ -0,0 +1,142 @@
+
+    return rows;
+}
+
+bool RowSparseDisk::load(std::istream &f) {
+    auto _f = dynamic_cast<mtg::common::IfstreamWithNameAndOffset *>(&f);
+    assert(_f);
+    try {
+        f.read(reinterpret_cast<char *>(&num_columns_), sizeof(uint64_t));
+
+        std::streampos boundary_start;
+        f.read(reinterpret_cast<char *>(&boundary_start), sizeof(boundary_start));
+
+        _f->SetOffset(static_cast<uint64_t>(f.tellg()));
+
+        int_vector_buffer_params.filename = _f->GetFName();
+        int_vector_buffer_params.offset = _f->GetOffset();
+
+        f.seekg(-sizeof(std::streampos), ios_base::end);
+
+        f.seekg(boundary_start, ios_base::beg);
+        boundary_.load(f);
+
+        num_rows_ = boundary_.num_set_bits();
+
+    } catch (...) {
+        return false;
+    }
+
+    return true;
+}
+
+void RowSparseDisk::serialize(std::ostream & /*f*/) const {
+    throw std::runtime_error(
+            "RowSparseDisk::serialize not supproted, use static variant instead");
+}
+
+
+void RowSparseDisk::serialize(const std::function<void(binmat::BinaryMatrix::RowCallback)> &call_rows,
+                              const std::string &filename,
+                              uint64_t num_cols,
+                              uint64_t num_set_bits,
+                              uint64_t num_rows) {
+    std::ofstream outstream(filename,
+                            std::ios::binary | std::ios::ate
+                                    | std::ios::in); // std::ios_base::in needed ...
+
+    if (!outstream.good())
+        throw std::ofstream::failure("Cannot write to file " + filename);
+
+    outstream.write(reinterpret_cast<char *>(&num_cols), sizeof(uint64_t));
+    auto boundary_start_pos = outstream.tellp();
+    // write "empty" boundary start
+    std::streampos boundary_start = 0;
+    outstream.write(reinterpret_cast<char *>(&boundary_start), sizeof(std::streampos));
+
+    const uint64_t iv_offs = outstream.tellp();
+    outstream.close();
+
+    auto outbuf = sdsl::int_vector_buffer<>(filename, std::ios::out | std::ios::binary,
+                                            1024 * 1024, sdsl::bits::hi(num_cols) + 1,
+                                            false, iv_offs);
+    // call_bits should be called for each argument that should be set (row boundaries)
+    auto call_bits = [&](const std::function<void(uint64_t)> &call_bit) {
+        uint64_t t = 0;
+        call_rows([&](const auto &row) {
+            uint64_t last = 0;
+            for (auto val : row) {
+                outbuf.push_back(val - last);
+                last = val;
+            }
+            call_bit((t += row.size() + 1) - 1);
+        });
+        outbuf.close();
+    };
+
+    bit_vector_small boundary(call_bits, num_rows + num_set_bits, num_rows);
+
+    outstream.open(filename,
+                   std::ios::in | std::ios::out | std::ios::binary | std::ios::ate
+                           | std::ios::in);
+    boundary_start = outstream.tellp();
```

Comment:
why does this need all of these flags?

Reply:
I needed to close the file before using `int_vector_buffer` because it will open and write to the same file. I will need to store the boundary vector, which position in the file is before `int_vector_buffer` content. When I reopen the file (after closing it via `int_vector_buffer`) I want to append the boundary vector (thus ios::ate). The problem is I cannot go back to the place where the boundary position is to be stored when opening with ios::ate unless I also specify ios::in.
Check this: https://stackoverflow.com/questions/47014463/ofstream-open-modes-ate-vs-app This is kinda confusing and counterintuitive that std::in must be used while it is out stream... I think std::out could be skipped but I decided to leave it here to be more verbose.
If you have any suggestions on how to do it better let me know!